pow: add the API to calculate next pow

vincenzopalazzo opened the pull request:
This is an initial work to support a helper function to calculate the next pow inside rust-bitcoin itself.

Fixes: https://github.com/rust-bitcoin/rust-bitcoin/issues/1997

This is out for early feedback on API design and also possible improvement that I did not see in the algorithm implementation

apoelstra commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +929,57 @@ impl kani::Arbitrary for U256 {
+    pow_target_timespan: u64,
+    no_pow_retargeting: bool,
+    get_block_by_heigh: F,
+) -> u32
```
In 003673d1cb0495b80fcea5e01333eba9594d8e20:

The closure can be changed to return a `BlockHeader` rather than a `Block`, which is much easier to deal with. Then if you provide `next_height` as a parameter you can collapse all the rest of the parameters:

* `last_height` is `next_height` minus one, doesn't need to be passed in
* `last_time` and `last_target` are in the blockheader for `last_height`, doesn't need to be passed in
* `adjustment_interval`, `pow_target_timespan`, `no_pow_retargeting` I think can all be set by a single`Network` argument to select between testnet/bitcoin/signet.

Next `get_block_by_height` needs to return a generic `Result` which should be propagated. Otherwise the user has no way to handle the case that they are missing blocks, requesting an invalid height, etc. If the user is sure their function can't fail they can provide `std::convert::Infallible` as the error type.

apoelstra replied:
Also, you have a typo `get_block_by_heigh` :).

apoelstra replied:
Also -- in the testnet case there is some rule where if there is no block for a long time the difficulty drops to 1, or something. It's fine to leave this as a TODO, but if we want to make this function generic then we should at least add a TODO comment.

Alternately, I'd be fine if we simply only supported Bitcoin ... then we could drop `adjustment_interval`, `pow_target_timespan` and `no_pow_retargeting` and just use fixed constants.

vincenzopalazzo replied:
Ok here is what I found after applying your suggestion

- where is the `next_height`? I am not sure I am using this as a procedure parameter, do I?
- The callback is already returning a Block Header, I just added the wrong name into it, or I am missing something?
- to remove `adjustment_interval`, `pow_target_timespan`, `no_pow_retargeting` I can use the `Params` struct?

>Also -- in the testnet case there is some rule where if there is no block for a long time the difficulty drops to 1, or something. It's fine to leave this as a TODO, but if we want to make this function generic then we should at least add a TODO comment.

If I want to support this in this PR what is the best approach, we return 1 if the network is testnet and there is no block generated for a long time?

apoelstra replied:
Oh, yeah, using the `Param`s struct is a good idea. And feel free to add fields to it if you need.

apoelstra replied:
As for `next_height`, yeah, oops, I'm not sure what I was thinking.

I would suggest renaming `last_height` to `current_height` though.

vincenzopalazzo replied:
Done!

apoelstra commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +929,57 @@ impl kani::Arbitrary for U256 {
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        adjusted_timespan = pow_target_timespan * 4;
+    }
```
In 003673d1cb0495b80fcea5e01333eba9594d8e20:

I think it'd be clearer to early return in these cases. For the returned `U256` you can use shifts rather than multiplies which will be much more efficient.

vincenzopalazzo replied:
I agree, but this optimisation is not done but the compiler itself? it should be easy on a type u64 to eeeeeeeeeeedo this kind of optimization, is it?

vincenzopalazzo replied:
BTW in here, we can return early in both If? or just for the multiplication?

apoelstra replied:
In both cases. If we would adjust difficulty by a factor 4 in either direction, the adjustment is clamped by the factor 4.

tcharding replied:
> In [003673d](https://github.com/rust-bitcoin/rust-bitcoin/commit/003673d1cb0495b80fcea5e01333eba9594d8e20):
> 
> I think it'd be clearer to early return in these cases. For the returned `U256` you can use shifts rather than multiplies which will be much more efficient.

I thought the compiler optimized this to a shift for us?

tcharding replied:
> I agree, but this optimisation is not done but the compiler itself? 

I thought this would have been optimized by the compiler too. Anyone know for sure?

Kixunil replied:
@tcharding I'm sure it's optimized for primitives (`u8`-`u128` and `usize`) not for our `U256`. IIRC `pow_target_timespan` is `u32`.

tcharding replied:
Even though the `U256` type is private? What if we used `target.0` to get at the `U256`?

tcharding replied:
Oh of course, `U256` is an "arbitrary" type not an actual integer type. I was thinking it was a simple wrapper `U256(<integer type>)`.

tcharding commented on bitcoin/src/pow.rs:
```diff
@@ -248,6 +250,22 @@ impl Target {
+    ///
+    /// The multiplication result along with a boolean indicating whether an arithmetic overflow
+    /// occurred. If an overflow occurred then the wrapped value is returned.
+    // mutagen false pos mul_u64: replace `|` with `^` (XOR is same as OR when combined with <<)
+    // mutagen false pos mul_u64: replace `|` with `^`
```
These comments are out put from `mutagen` runs on the other `mul_u64` function, we shouldn't duplicate them :)

vincenzopalazzo replied:
Yeah, I just copied and pasted it sorry. I try to ask for clarification in IRC about this mutagen usage. Can you clarify why these feature flags are used?

tcharding replied:
[`mutagen`](https://github.com/llogiq/mutagen) is a mutation testing program that modifies the codebase then runs our unit tests to see if the mutation caused the tests to fail. If the tests do not fail the mutation is said to have survived - indicating we need more unit tests. There are some false positives and these comments are the best way we could come up with to aid future devs running the tests to realize they are false positives.

vincenzopalazzo replied:
Oh that it is nice to know, thanks

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+/// TODO: add some docs here
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
+    consensius_params: Params,
```
nit: spelling

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+pub fn calculate_next_work_required<F>(
+    consensius_params: Params,
+    last_height: Height,
+    get_block_header_by_heigh: F,
```
nit: spelling

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+    let current_height: u64 = last_height.to_consensus_u32() as u64;
+
+    if (current_height % adjustment_interval.min(1)) != 0 {
+        return Ok(last_adjustment_block.target().to_compact_lossy().to_consensus());
```
If adjustment_interval is 2016 and current height is 2019, then the "next work required" here would be the target used between block 0~2015... that doesn't seem right.

vincenzopalazzo replied:
I use the same logic https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L19

Can you be more precise in what I can do to make this look right? thanks

junderw replied:
`pindexLast` in bitcoin core is the equivalent of the block at `current_height` in your code.

However, `last_adjustment_block` is the block at `current_height - 2015`

The variable names in bitcoin core's code are confusing, and your code also uses confusing names that are trying to mimic core's names.

Perhaps you should rename all the variables to make things easier to follow.

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+    }
+
+    if no_pow_retargeting {
+        return Ok(last_adjustment_block.bits.to_consensus());
```
This also doesn't seem right.

vincenzopalazzo replied:
I use the same logic https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L52

Can you be more precise in what I can do to make this look right? thanks

junderw replied:
See above comment: https://github.com/rust-bitcoin/rust-bitcoin/pull/2090/files#r1346210181

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+    }
+
+    let last_time = last_adjustment_block.time;
+    let actual_timespan = (last_time - last_adjustment_time) as u64;
```
This will always be 0. You are subtracting the same value with itself.

vincenzopalazzo replied:
Good catch, I think I need to leave the current_time as procedure input.

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+        adjusted_timespan = pow_target_timespan / 4;
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        return Ok(U256::from(pow_target_timespan).mul_u64(4).0.bits());
```
Why are you returning "the number of seconds in 2 weeks" (for mainnet) * 4?

Doesn't this fn need to return the 32 bit target shortened notation (like the other early returns)?

vincenzopalazzo replied:
Yeah you are right, I can use `Ok(pow_target_timespan * 4)` and cast to `u32`

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,81 @@ impl kani::Arbitrary for U256 {
+    let target = last_adjustment_block.target();
+    let (mut target, overflow) = target.mul_u64(adjusted_timespan);
+    // FIXME: improve the API of the target.
+    target = Target(target.0 / U256::from(pow_target_timespan));
```
Why are you dividing the target by "the number of seconds in 2 weeks"?

vincenzopalazzo replied:
For this reason https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp#L66

I am missing something?

junderw replied:
Copying the bad habits of the original source code is not a requirement. It's much more difficult in C++ to change the name of variables a lot, so they recycle the same variable a lot. This is not needed in Rust.

`target` is no longer a target, but it's a stand-in for a `U256` here that represents the mid-state of a calculation.

You are basically multiplying the target by the ratio of `adjusted_timespan / pow_target_timespan`

This calculation is ok, but the variable names should change between mid-state and final result.

Re-using the `target` variable is confusing.

junderw reviewed (changes requested):
The calculations in this function are unrelated to the calculation of a new target.

Please fix the calculations, and if possible the spelling nits too.

vincenzopalazzo commented:
I've rebased and pushed the changes that I understood from the review. I've also added the testnet work as future work, as discussed with @apoelstra.

---

Regarding some of the feedback that @junderw provided, I couldn't find a solution. I was only told that my approach was incorrect, but I'm unsure how to modify my code to address the identified issues.

Could you please provide a clearer review to guide me in fulfilling your requirements?

Thank you.

junderw commented:
> Could you please provide a clearer review to guide me in fulfilling your requirements?

The requirements aren't special. Just "perform correct calculations."

I don't understand how much clearer I can be without re-writing your code completely.

Adding tests might help prove your calculations are right. You can even take some values from mainnet blocks and plug them in and make sure you get the same results.

junderw commented:
Please see https://github.com/vincenzopalazzo/rust-bitcoin/pull/1 for fixes.

junderw commented:
Found a bug in the Work constants https://github.com/rust-bitcoin/rust-bitcoin/issues/2103

vincenzopalazzo commented:
>This calculation is ok, but the variable names should change between mid-state and final result.

You are right, I get completely confused! now review your code I get where I was going. Thanks for doing that

tcharding commented on bitcoin/src/pow.rs:
```diff
@@ -248,6 +249,20 @@ impl Target {
     pub fn max_difficulty_transition_threshold(&self) -> Self { Self(self.0 << 2) }
+
+    /// Returns the least number of bits needed to represent the number.
+    pub fn bits(&self) -> u32 { self.0.bits() }
```
Why add this?

vincenzopalazzo replied:
This is no longer needed, thanks

tcharding commented on bitcoin/src/pow.rs:
```diff
@@ -248,6 +249,20 @@ impl Target {
+    ///
+    /// The multiplication result along with a boolean indicating whether an arithmetic overflow
+    /// occurred. If an overflow occurred then the wrapped value is returned.
+    pub fn mul_u64(self, rhs: u64) -> (Target, bool) {
```
This should be private, right? Isn't the whole motivation for adding the `calculate_next_work_required` function so we don't expose arbitrary math operations on the target/work types?

vincenzopalazzo replied:
Correct

tcharding commented on bitcoin/src/pow.rs:
```diff
@@ -400,7 +415,6 @@ impl U256 {
     }
 
     /// Returns the least number of bits needed to represent the number.
-    #[cfg_attr(all(test, mutate), mutate)]
```
These shouldn't be removed please, they are used when mutation testing.

vincenzopalazzo replied:
Ops, I think this came from @junderw PR, I merged and forgot to add them, thanks for the catch

tcharding commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,116 @@ impl kani::Arbitrary for U256 {
+/// Occurse when there is an error while get a new block header
+/// from an external source.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum GetBlockHeaderError {
+    /// Block Header not found
+    NotFound,
+    /// Generic error with custom message
+    Generic(String),
+}
```
Perhaps this would be better
```rust
/// Error from `get_block_header_by_height` callback.
///
/// Occurse when there is an error while get a new block header
/// from an external source.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum GetBlockHeaderError<E> {
    /// Block Header not found
    NotFound,
    /// Custom error.
    Custom(E),
}

impl<E: fmt::Display> fmt::Display for GetBlockHeaderError<E> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use GetBlockHeaderError::*;

        match *self {
            Custom(ref e) => write_err!(f, "custom error"; e),
            NotFound => f.write_str("block header not found"),
        }
    }
}

#[cfg(feature = "std")]
impl<E: std::error::Error + 'static> std::error::Error for GetBlockHeaderError<E> {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        use GetBlockHeaderError::*;

        match *self {
            Custom(ref e) => Some(e),
            NotFound => None,
        }
    }
}
```
Can you put all the error stuff _below_ where it is used i.e., below `calculate_next_work_required`

vincenzopalazzo replied:
yes, I was thinking about this yesterday, thanks for writing the code. Apply this

tcharding reviewed (commented):
I did not review the logic in the `calculate_next_work_required` function, to do so requires me to understand the Core code since this mirrors it. It would be really nice if that was not the case, eg. what do the div by 4 and mul by 4s mean?

You can ignore me if you like, but we will need to get 2 reviewers to understand the logic.

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,116 @@ impl kani::Arbitrary for U256 {
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    get_block_header_by_height: F,
+) -> Result<u32, GetBlockHeaderError>
+where
+    F: Fn(u32) -> Result<Header, GetBlockHeaderError>,
```
FnMut is less restrictive, and since we are able to call it mutably, we should... unfortunately, since it can possibly be called multiple times on testnet, we can't use FnOnce.

```suggestion
    mut get_block_header_by_height: F,
) -> Result<u32, GetBlockHeaderError>
where
    F: FnMut(u32) -> Result<Header, GetBlockHeaderError>,
```

vincenzopalazzo replied:
I do not see why we should not do this, applying

junderw commented on bitcoin/src/pow.rs:
```diff
@@ -913,6 +926,116 @@ impl kani::Arbitrary for U256 {
+            Self::Generic(error) => error,
+            Self::NotFound => "Block header not found",
+        };
+        write!(f, "{msg}")
```
```suggestion
        write!(f, "{}", msg)
```

junderw replied:
This is the reason CI is breaking. (1.48.0 Rust can't embed variables like this)

vincenzopalazzo replied:
ops I forget this

tcharding commented:
Is the  4 something to do with this: "A single retarget never changes the target by more than a factor of 4 either way to prevent large changes in difficulty."?

ref: https://en.bitcoin.it/wiki/Target

junderw commented:
> Is the 4 something to do with this: "A single retarget never changes the target by more than a factor of 4 either way to prevent large changes in difficulty."?
> 
> ref: https://en.bitcoin.it/wiki/Target

Yes. The largest target increase is << 2 and the smallest target decrease is >> 2. (*4 and /4)

bit shifts might be more efficient, but less clear when reading. (Core uses * and /)

junderw commented:
#2103 is a show stopper though... since we need to compare the final Target to the consensus Params pow largest target.

Perhaps we should change the type of the Param value to Target just like Core does. Then the Work/Target conversion bug wouldn't be a show stopper here.

vincenzopalazzo commented:
>I did not review the logic in the calculate_next_work_required function, to do so requires me to understand the Core code since this mirrors it. It would be really nice if that was not the case, eg. what do the div by 4 and mul by 4s mean?

Sorry about that, it looks difficult also to me! I did not find good documentation for writing it from scratch, maybe I am missing some link

I found this very helpful https://bitcoin.stackexchange.com/a/81878/91194 but I think @junderw got the understanding correctly more than me 😄 


>You can ignore me if you like, but we will need to get 2 reviewers to understand the logic.

Why I should do that? 😸 

>Perhaps we should change the type of the Param value to Target just like Core does. Then the Work/Target conversion bug wouldn't be a show stopper here.

This sounds like a good idea but introduces a breaking change and I would like to avoid without due to a lot of applications are just now keeping up with bitcoin 30. Maybe we can add a new field and then deprecate the old one? An when the time comes we drop it?

I am not sure what are the deprecation policy there btw

vincenzopalazzo commented:
What about this @junderw https://github.com/rust-bitcoin/rust-bitcoin/pull/2090/commits/e8cfd4221fb4407f3a53f2b524ca8fcad1f199b8

I do not know if the value of Targed is done correctly, if my solution will be an option I can think a little bit harder 🦺 and fix it

junderw commented on bitcoin/src/consensus/params.rs:
```diff
@@ -61,6 +71,7 @@ impl Params {
                 rule_change_activation_threshold: 1916, // 95%
                 miner_confirmation_window: 2016,
                 pow_limit: Work::MAINNET_MIN,
+                pow_limit_target: Target::from_le_bytes(Work::MAINNET_MIN.to_le_bytes()),
```
This is wrong. You need to get the values from Bitcoin Core.

vincenzopalazzo replied:
>This is wrong. You need to get the values from Bitcoin Core.

I was surprize by the opposite 😄 as I mentioned in my previous message I do not know if the value of Targed is done correctly. I would like to discuss the solution to add it without breaking change

tcharding replied:
Since we are uncovering bugs I vote for just doing a breaking change and getting the whole thing right.

In general we use `deprecated` when possible and give folk one version before we remove it. This is mainly to try and be gentle on users but since we are pre-1.0 and are trying _really_ hard to stabalize getting a good API takes precedence over deprecating and/or keeping kruft around.

vincenzopalazzo replied:
Make sense, I will finish the patch today. I spent some time understanding the different between the Work type and the Target type before

vincenzopalazzo commented:
Ok waiting for https://github.com/rust-bitcoin/rust-bitcoin/pull/2107 and I will drop the fixup commit then

Thanks to help in this bug, I am quite lost with this Work/Target type conversion

vincenzopalazzo commented:
Trivial rebase on master to fix the target bug.

tcharding commented:
Hey mate, I spent the afternoon playing with this PR. The only way I understand hard things is to code them myself so I re-wrote a new _buggy_ version. I can write a review here but it is probably easier for you to just take a look and grab anything you like and leave out anything you don't. Then I'll re-review again.

https://github.com/tcharding/rust-bitcoin/blob/606bfe93d0268f7f29c3234ee4072b8c39ae79ee/bitcoin/src/pow.rs#L955

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+    current_height: Height,
+    current_header: Header,
+    new_block_timestamp: u32,
+    mut get_block_header_by_height: F,
```
I think many projects will desire `async` support here, can we get rid of the closure or something? IIUC this is just pulling all the blocks from the retarget period so we could just accept `IntoIterator<Item=I>, I: Borrow<Header>`. People would then just fetch all the blocks first and compute the difficulty afterwards. (I know it's a bit wasteful but I don't have a better idea that doesn't involve confusing API workarounds.) This also makes tests much easier.

Also if we do end up with closure we should use `Height` in the parameter.

apoelstra replied:
We don't need all the blocks; we only need the starting and ending blocks.

apoelstra replied:
So I guess there's an even easier soluton where we just take those headers directly as arguments :).

tcharding replied:
If we only take the first and last block [header] we put the burden of doing this loop on the user

https://github.com/bitcoin/bitcoin/blob/82ea4e787c791acbc85fd3043dd6bae038cba4f2/src/pow.cpp#L32

FTR I don't actually understand what the checks are for in this loop.

tcharding replied:
(And they have to understand that this loop is only required for testnet, so no one is going to do it.)

vincenzopalazzo replied:
>I think many projects will desire async support here, can we get rid of the closure or something?

Does the async support require also an async caller? otherwise, it is just a blocking call, right?

But I also like the idea of passing it as an input parameter, and the @tcharding is a good point.

This is the era of choice, we must support async or we do not want it at the API level right now? I leave you the fun of choose 😄 because I am bias

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+        return Ok(current_header.bits.to_consensus());
+    }
+
+    if no_pow_retargeting {
```
This could go to around the beginning of the function to avoid costly computation that gets thrown away.

vincenzopalazzo replied:
Good point

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+    }
+    if actual_timespan > pow_target_timespan * 4 {
+        adjusted_timespan = pow_target_timespan * 4;
+    }
```
This whole thing could be just `let adjusted_timespan = actual_timespan.clamp(pow_target / 4, pow_target *4)`.

vincenzopalazzo replied:
Thanks!

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+    let pow_limit_target = consensus_params.pow_limit;
+    if target > pow_limit_target {
+        target = pow_limit_target;
+    }
```
`let target = target.min(consensus_params.pow_limit);`

vincenzopalazzo replied:
Thanks!

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+        target = pow_limit_target;
+    }
+
+    Ok(target.to_compact_lossy().to_consensus())
```
Why the last `to_consensus` call (and returning `u32`)? Strong types are generally better.

vincenzopalazzo replied:
I do not remember the why, but you are right

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+///
+/// Return a error if the callback provide from the caller fails.
+#[cfg_attr(all(test, mutate), mutate)]
+pub fn calculate_next_work_required<F>(
```
This function says `work` but actually returns target, those are not the same thing so either the name is wrong or the returned value. (I guess it's the name.)

tcharding replied:
Its because the name is copied from Core, I agree, we should return a `CompactTarget` and call it `calculate_next_target` or some such name.

vincenzopalazzo replied:
Done! thanks

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+
+    if (current_height + 1) % adjustment_interval != 0 {
+        if allow_min_difficulty_blocks {
+            let pow_limit = consensus_params.pow_limit.to_compact_lossy().to_consensus();
```
I'd put this inside `if` to more obviously avoid costly computation.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+            } else {
+                use hashes::Hash;
+                let mut header = current_header;
+                let mut height = current_height;
```
Crap, this is confusing, looks like duplicated line! :D (there's no solution, I'm just ranting)

tcharding replied:
ha, I tried twice to make this nicer and couldn't come up with anything.

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+            if new_block_timestamp > current_header.time + pow_target_spacing * 2 {
+                return Ok(pow_limit);
+            } else {
+                use hashes::Hash;
```
OT: we really should have `all_zeros` as an inherent method. I think I did that in split hash triat PR.

tcharding replied:
Done in #2496

Kixunil commented on bitcoin/src/pow.rs:
```diff
@@ -923,6 +935,110 @@ impl kani::Arbitrary for U256 {
+                let mut height = current_height;
+                while header.prev_blockhash != BlockHash::all_zeros()
+                    && height % adjustment_interval != 0
+                    && header.bits.to_consensus() == pow_limit
```
You're converting everything to consensus and then comparing. You could (and most likely should) just not convert and compare the strong types instead. In fact, if there's any reason not to something is probably horribly wrong.

apoelstra replied:
I don't think we have comparison opcodes on `CompactTarget`. If we do we are planning to get rid of them. See https://github.com/rust-bitcoin/rust-bitcoin/issues/2110. (I see that you commented there that we should have explict named methods for the different comparison methods. Ok, but they don't exist right now.)

tcharding replied:
FWIW, the pow stuff is on the top of my list at the moment, I started on it yesterday.

Kixunil replied:
Oh, crap, so we have an actual use case for comparison, kinda undermining the point of #2110, but it's still obscure. It's also super weird that Core compares those. Makes me quite intrigued to figure out WTF is going on.

Using `Eq` now could be a nice way to not forget to change the code (presumably to use explicit method) when #2110 is resolved.

apoelstra replied:
Well, yes, we have a use case *within PoW computation*. But there's lots of stuff we do in that computation that we don't necessarily want to expose by itself.

tcharding commented:
@Kixunil suggested new functions `min`, `clamp` help a lot, IMO. The code I wrote yesterday _still_ looks a lot like C++.

tcharding commented:
copied from https://github.com/lightningdevkit/rust-lightning/issues/2124#issuecomment-1799975693

>> So you guys need https://github.com/rust-bitcoin/rust-bitcoin/pull/2090 merged and backported to 0.30.1 as well as 0.31.1, right?

> Is ready for review, this PR is blocking me on nakamoto, but I think I will have a followup regarding the API.

I don't see any force pushes in the last 5 days, is the github UI glitching?

vincenzopalazzo commented:
>suggested new functions min, clamp help a lot, IMO. The code I wrote yesterday still looks a lot like C++.

What is wrong with being like C++? I do not see anything wrong with C++ style. However, the bad thing about the C++ version is that there are more than 4 people trying to understand the core code, so the only thing that would be good to improve is the documentation and the code, other black magic rust is just making the core difficult to read when the time to say "I wrote yesterday still looks a lot like Rust." will come.

I will go through the review over the weekend

Kixunil commented:
> What is wrong with being like C++? I do not see anything wrong with C++ style.

There's a reason we use quite different style in Rust. Aside from many bugs, Rust is much more readable.

Min, max a and clamp are not black magic.

vincenzopalazzo commented:
>Min, max a and clamp are not black magic.

This is also possible in C++ with `std::min` and `std::max` but other than that (I did not write the code that you reviewed, the commit is Co-Developed if you noted) I do not see why we should change the structure of the code that we do not understand well enough to change and simplify from the core version.

Other than using the `min`, `man` and `clamp` simplification I do not expect to see this function structure different from the core version, except if there is someone who will step up and take ownership of the code by also fulfilling the needs of the user libraries like nakamoto, ldk and friends

vincenzopalazzo commented:
>I don't see any force pushes in the last 5 days, is the github UI glitching?

Sorry @tcharding I got overwhelmed by the rust-bitcoin notification and I lost it inside my notification view. You should see the force push now

Kixunil commented:
I, and likely most Rustceans, find those methods more readable. That's a good enough reason already. But it's not something I would block a PR for. But it's possible we'll have to change it anyway if a future version of clippy has a lint for this.

vincenzopalazzo commented:
```
    Checking secp256k1 v0.28.0
error[E0658]: use of unstable library feature 'clamp'
    --> bitcoin/src/pow.rs:1018:45
     |
1018 |     let adjusted_timespan = actual_timespan.clamp(pow_target_timespan / 4, pow_target_timespan * 4);
     |                                             ^^^^^
     |
     = note: see issue #44095 <https://github.com/rust-lang/rust/issues/44095> for more information

error: aborting due to previous error

For more information about this error, try `rustc --explain E0658`.
error: could not compile `bitcoin`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error[E0658]: use of unstable library feature 'clamp'
    --> bitcoin/src/pow.rs:1018:45
     |
1018 |     let adjusted_timespan = actual_timespan.clamp(pow_target_timespan / 4, pow_target_timespan * 4);
     |                                             ^^^^^
     |
     = note: see issue #44095 <https://github.com/rust-lang/rust/issues/44095> for more information

error: aborting due to previous error
```

cc @tcharding @Kixunil

Kixunil commented:
Oh, it has MSRV 0.50, sorry about that. Skip that one and consider adding `// TODO: update to clamp in MSRV >= 0.50`

tcharding commented:
> I do not see why we should change the structure of the code that we do not understand well enough to change and simplify from the core version.

Who is the second "we" referring to? After spending time on this last week I am more than comfortable to say I understand the core code and can re-write it in Rust style.

> except if there is someone who will step up and take ownership of the code by also fulfilling the needs of the user libraries like nakamoto, ldk and friends

I can take ownership of this, I was intentionally not doing so because I didn't want to steam roll over you since you got here first. If you don't mind how this gets done and just want to see the functionality in `rust-bitcoin` then lets close this PR and I'll push it in through #2180. I'm super keen to get this in because LDK is waiting on it so they can upgrade.

If there is any other functionality missing that nakamoto needs please do say so, I'm working me hardest to support folk these days. I have not personally looked at the nakamoto codebase though.

To clarify, if you leave this open I'll assume you want to do the work and will review and wait for you. If you close this PR I'll push forward with #2180.

junderw commented:
I think we should close in favor of 2180. I am willing to review this PR again though, if it needs reviewing.

Kixunil commented:
I think the sync API is much more concerning than any code style around min/max.

vincenzopalazzo commented:
@tcharding up to you, I really do not care who is the author of the PR or how the code will be written more rusty or not Cpp style. I care just that if I have some bugs the code is clear that I can read it without cry

The code that will be merged will have no API user, so we do not know if with this API I will be able to sync nakamoto with the network and fully validate the chain or if the user of the API needs some different UX. So my priority is to have this functionality so I can continue to do my work and provide more tests about the code.

When we have this code merged then I need to way the wallet user to update to the new version of rust bitcoin and then I can test nakamoto. So this sounds a work that will take months.

>I think the sync API is much more concerning than any code style around min/max.

We can also think of moving the function in an async style due to that in the rust compiler we are doing some work to make the async programming usable. However, this is not good for people who are using not standard async IO (e.g nakamoto) or low-level API with mio and I do not like forcing this async style without having some concrete user. e.g: If nakamoto is going to use this API we do not need an async function

Kixunil commented:
Maybe we can modify the API to be completion based? I know it can have performance issues but if it's done right it might be fine. Like what I did with `push_decode`.

tcharding commented:
> @tcharding up to you, I really do not care who is the author of the PR 

Can you close this if you are ok for me to work on #2180, I don't want to close your PR.

> or how the code will be written more rusty or not Cpp style. I care just that if I have some bugs the code is clear that I can read it without cry

Don't worry, writing code that is brain-dead simple to read is my thing :)
